src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
+
+        let tree_height = tree_height(non_leaf_nodes_size + leaf_nodes_size);
+
+        let hash_of_empty: TwoToOneDigest<P> = P::TwoToOneHash::evaluate_both(
+            two_to_one_hash_param,
+            &vec![0u8; P::TwoToOneHash::LEFT_INPUT_SIZE_BITS / 8],
+            &vec![0u8; P::TwoToOneHash::RIGHT_INPUT_SIZE_BITS / 8],
+        )?;
+
+        // initialize the merkle tree as array of nodes in level order
+        let mut non_leaf_nodes: Vec<TwoToOneDigest<P>> = (0..non_leaf_nodes_size)
+            .map(|_| hash_of_empty.clone())
+            .collect();
+        let mut leaf_nodes: Vec<LeafDigest<P>> = Vec::with_capacity(leaf_nodes_size);
+
+        // Compute the starting indices for each non-leaf level of the tree
         let mut index = 0;
-        let mut level_indices = Vec::with_capacity(tree_height);
-        for _ in 0..tree_height {
+        let mut level_indices = Vec::with_capacity(tree_height - 1);
+        for _ in 0..(tree_height - 1) {
             level_indices.push(index);
             index = left_child(index);
         }
 
-        // Compute and store the hash values for each leaf.
-        let last_level_index = level_indices.pop().unwrap_or(0);
-        let mut buffer = [0u8; 128];
-        for (i, leaf) in leaves.iter().enumerate() {
-            tree[last_level_index + i] = hash_leaf::<P::H, _>(&parameters, leaf, &mut buffer)?;
+        // compute and store hash values for each leaf
+        for leaf in leaves.iter() {
+            leaf_nodes.push(P::LeafHash::evaluate(
+                leaf_hash_param,
+                &ark_ff::to_bytes!(leaf)?,
+            )?)
+        }
+
+        // compute the hash values for the non-leaf bottom layer
+        {
+            let start_index = level_indices.pop().unwrap();
+            let upper_bound = left_child(start_index);
+            for current_index in start_index..upper_bound {
+                let left_leaf_index = left_child(current_index) - upper_bound;
+                let right_leaf_index = right_child(current_index) - upper_bound;
+                // compute hash
+                let left_bytes = ark_ff::to_bytes!(&leaf_nodes[left_leaf_index])?;
+                let right_bytes = ark_ff::to_bytes!(&leaf_nodes[right_leaf_index])?;
+                non_leaf_nodes[current_index] = P::TwoToOneHash::evaluate_both(
+                    &two_to_one_hash_param,
+                    &left_bytes,
+                    &right_bytes,
+                )?
+            }
         }
 
-        // Compute the hash values for every node in the tree.
-        let mut upper_bound = last_level_index;
-        let mut buffer = [0u8; 128];
+        // compute the hash values for nodes in every other layer in the tree
         level_indices.reverse();
         for &start_index in &level_indices {
-            // Iterate over the current level.
+            let upper_bound = left_child(start_index); // The exclusive index upper bound for this layer
```

Comment:
```suggestion
            // The exclusive index upper bound for this layer
            let upper_bound = left_child(start_index);
```
Also not sure what exclusive index means here

Reply:
it means "The layer beginning `start_index` ends at `upper_bound` (exclusive)."

Reply:
resolved